frontend/common/src/helpers/router.tsx

```diff
@@ -87,35 +98,84 @@ export const pushToStateThenNavigate = (url: string, state: State): void => {
 };
 
 export interface RouterResult {
-  component: ReactElement;
+  component?: ReactElement;
   redirect?: string;
+  authorizedRoles?: Array<Role>;
 }
 
 export const useRouter = (
   routes: Routes<RouterResult>,
   missingRouteComponent: ReactElement,
+  notAuthorizedComponent: ReactElement,
+  isLoggedIn: boolean,
+  loggedInUserRoles: PossibleUserRoles,
 ): React.ReactElement | null => {
   const location = useLocation();
   const router = useMemo(() => new UniversalRouter(routes), [routes]);
   const [component, setComponent] = useState<React.ReactElement | null>(null);
   const pathName = location.pathname;
+  const apiRoutesRef = useRef(useApiRoutes()); // I'm not not sure why this function is unstable on its own!
```

Comment:
Would love it if anyone could help me understand why this function was unstable.  If it is added directly to the useEffect dependencies it causes render loops.

Reply:
There is an unnecessary nesting of functions. I'm not sure why this causes infinite re-renders but it certainly adds complexity. Since `apiRoutes` is static and takes no arguments, its safer to declare it as an object rather than a function that returns an object. Try this:

```tsx
import path from "path-browserify";

export interface ApiRoutes {
  login: (from?: string, locale?: string) => string;
  refreshAccessToken: () => string;
}

const apiRoot = (): string => "/";

const apiRoutes = {
  login: (from?: string, locale?: string): string => {
    const searchTerms: string[] = [];
    if (from) searchTerms.push(`from=${encodeURI(from)}`);
    if (locale) searchTerms.push(`locale=${encodeURI(locale)}`);
    const searchString = searchTerms.join("&");

    const url =
      path.join(apiRoot(), "login") + (searchString ? `?${searchString}` : "");

    return url;
  },
  refreshAccessToken: (): string => path.join(apiRoot(), "refresh"),
};

/**
 * A hook version of loginRoutes which gets the locale from the intl context.
 * @returns LoginRoutes
 */
export const useApiRoutes = (): ApiRoutes => {
  return apiRoutes;
};

```

Reply:
Yep, that was totally it.  Yer a wizard, @esizer!  :mage:

Reply:
For a bit more context, while it may be unnecessary for api routes, the talentSearchRoutes and adminRoutes use hooks to get the current locale from react-intl context, and inject it into the routes.